torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -20,50 +23,265 @@ std::unique_ptr<RRefContext>& RRefContext::getInstance() {
 }
 
-void RRefContext::addFork(const at::IValue& value) {
-  auto rfd = RRefForkData::fromIValue(value);
-  AT_ASSERT(
-      rfd.ownerId_ == getWorkerId(),
-      "RRef user should never receive fork notification.");
+template std::shared_ptr<UserRRef<IValue>> RRefContext::createUserRRef<IValue>(
+    worker_id_t ownerId,
+    const RRefId& rrefId,
+    const ForkId& forkId);
+
+template std::shared_ptr<UserRRef<py::object>> RRefContext::createUserRRef<
+    py::object>(
+    worker_id_t ownerId,
+    const RRefId& rrefId,
+    const ForkId& forkId);
+
+template <typename T>
+std::shared_ptr<RRef> RRefContext::getOrCreateRRef(const RRefForkData& rfd) {
+  auto& ownerId = rfd.ownerId_;
+  auto& rrefId = rfd.rrefId_;
+  auto& forkId = rfd.forkId_;
+  if (ownerId == getWorkerId()) {
+    return getOrCreateOwnerRRef<T>(rrefId);
+  } else {
+    return createUserRRef<T>(ownerId, rrefId, forkId);
+  }
+}
+
+template std::shared_ptr<RRef> RRefContext::getOrCreateRRef<IValue>(
+    const RRefForkData& rfd);
+
+template std::shared_ptr<RRef> RRefContext::getOrCreateRRef<py::object>(
+    const RRefForkData& rfd);
+
+template <typename T>
+std::shared_ptr<OwnerRRef<T>> RRefContext::getOrCreateOwnerRRef(
+    const RRefId& rrefId) {
   std::lock_guard<std::mutex> lock(mutex_);
-  auto& rrefForks = forks_[rfd.rrefId_];
-  AT_ASSERT(
-      rrefForks.find(rfd.forkId_) == rrefForks.end(),
-      "Got fork notification twice on the same RRef ",
-      rfd.rrefId_);
-  rrefForks.insert(rfd.forkId_);
+  const auto iter = owners_.find(rrefId);
+  if (iter == owners_.end()) {
+    // Scenario (1) the first time this owner knows about this RRef
+    // Scenario (2) This owner is also the creator.
+    //
+    // NB: cannot use make_shared here as the constructor of OwnerRRef is
+    // private.
+    auto rref =
+        std::shared_ptr<OwnerRRef<T>>(new OwnerRRef<T>(getWorkerId(), rrefId));
+    owners_[rref->rrefId()] = rref;
+    return rref;
+
+  } else {
+    // Scenario (3) retrieving an existing RRef
+    return std::static_pointer_cast<OwnerRRef<T>>(iter->second);
+  }
+}
+
+template std::shared_ptr<OwnerRRef<IValue>> RRefContext::getOrCreateOwnerRRef<
+    IValue>(const RRefId& rrefId);
+
+template std::shared_ptr<OwnerRRef<py::object>> RRefContext::
+    getOrCreateOwnerRRef<py::object>(const RRefId& rrefId);
+
+RRefForkData RRefContext::prepareChildFork(const std::shared_ptr<RRef>& rref) {
+  auto rfd = rref->fork();
+  if (rref->isOwner()) {
+    addForkOfOwner(rfd.rrefId_, rfd.forkId_);
+  } else {
+    if (rref->isPyObj()) {
+      addPendingChild(
+          rfd.forkId_, std::static_pointer_cast<UserRRef<py::object>>(rref));
+    } else {
+      addPendingChild(
+          rfd.forkId_, std::static_pointer_cast<UserRRef<IValue>>(rref));
+    }
+  }
+  return rfd;
+}
+
+void RRefContext::notifyOwnerAndParentOfFork(
+    const ForkId& forkId,
+    worker_id_t parent,
+    const std::shared_ptr<RRef>& rref) {
+  if (parent != rref->owner()) {
```

Comment:
In the current code, if the parent is the owner, the parent will directly add the fork into `forks_` which will be deleted when the child exits. I can see how this could cause problems on errors, i.e., the fork is added, but the delete of the fork will never come. Let me add an ack here.

Reply:
I am referring to the case where the rpc/remote message got dropped. We will not retry the message, but the current code would have already incremented the reference count if the caller is the owner. It might be better to increment the reference count when the owner gots the ACK.

Reply:
@gqchen I thought about this again. You are right, the current implementation is fine, and it is consistent with user->user fork. The current implementation does the following for the case where owner B shares to user C:

1. B adds the fork into `forks_` to the `OwnerRRef` alive.
2. B sends the message (rpc or remote) to the callee user C.
3. The user C creates a `UserRRef` and acks the owner. 

In the above case, the fork was never in a pending states on the owner. Even if we only add the `OwnerRRef` into `forks_` when the ack arrives, we still have to add it in some other map (e.g., `pendingChildren_`) to keep the `OwnerRRef` alive before getting the ACK. So, it does not make any difference but only adds complexity. 

When we have error handling, if B didn't receive an ACK and timed out, it should drop the fork, and there will be no leak. I will keep the protocol as is and only apply the code simplification suggestions.